nab/detectors/htmjava/src/main/java/nab/detectors/htmjava/HTMModel.java

```diff
@@ -0,0 +1,266 @@
+package nab.detectors.htmjava;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.joda.time.DateTimeZone;
+import org.numenta.nupic.Connections;
+import org.numenta.nupic.Parameters;
+import org.numenta.nupic.Parameters.KEY;
+import org.numenta.nupic.algorithms.Anomaly;
+import org.numenta.nupic.algorithms.SpatialPooler;
+import org.numenta.nupic.algorithms.TemporalMemory;
+import org.numenta.nupic.network.Layer;
+import org.numenta.nupic.network.Network;
+import org.numenta.nupic.network.Region;
+import org.numenta.nupic.network.sensor.ObservableSensor;
+import org.numenta.nupic.network.sensor.Publisher;
+import org.numenta.nupic.network.sensor.Sensor;
+import org.numenta.nupic.network.sensor.SensorParams;
+import org.numenta.nupic.util.Tuple;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+
+public class HTMModel {
+    protected static final Logger LOGGER = LoggerFactory.getLogger(HTMModel.class);
+
+    private Network network;
+
+    private Publisher publisher;
+
+    /**
+     * Create HTM Model to be used by NAB
+     * @param modelParams OPF Model parameters to parameters from
+     */
+    public HTMModel(JsonNode modelParams) {
+        if(LOGGER.isDebugEnabled()) {
+            LOGGER.debug("OPF Model Params:" + modelParams.toString());
+        }
+
+        // Create Sensor publisher to push NAB input data to network
+        publisher = Publisher.builder()
+                .addHeader("timestamp,value")
+                .addHeader("datetime,float")
+                .addHeader("T,B")
+                .build();
+
+        // Get updated model parameters
+        Parameters parameters = getModelParameters(modelParams);
+        if(LOGGER.isDebugEnabled()) {
+            LOGGER.debug("HTM java Parameters:" + parameters.toString());
+        }
+
+        // Create NAB Network
+        network = Network.create("NAB Network", parameters)
+                .add(Network.createRegion("NAB Region")
+                    .add(Network.createLayer("TemporalMemory", parameters)
+                            .add(Anomaly.create())
+                            .add(new TemporalMemory()))
+                    .add(Network.createLayer("SpatialPooler", parameters)
+                            .add(new SpatialPooler()))
+                    .add(Network.createLayer("Sensor", parameters)
+                            .add(Sensor.create(ObservableSensor::create,
+                                    SensorParams.create(SensorParams.Keys::obs, "Manual Input", publisher))))
+                    .connect("TemporalMemory", "SpatialPooler")
+                    .connect("SpatialPooler", "Sensor"));
+
```

Comment:
While adding the algorithms to their own "Layers" is totally legitimate, I wouldn't do it in production unless it served a particular purpose because it will run a bit slower. An alternative would be: (no need to connect anything here).

```Java
network = Network.create("NAB Network", parameters)
                 .add(Network.createRegion("NAB Region")
                    .add(Network.createLayer("SheinkmanLayer", parameters)
                             .add(Anomaly.create())
                             .add(new TemporalMemory())
                             .add(new SpatialPooler())
                             .add(Sensor.create(ObservableSensor::create,
                                     SensorParams.create(SensorParams.Keys::obs, "Manual Input", publisher))));